packages/flutter_tools/lib/src/dart_pub_json_formatter.dart

```diff
@@ -16,31 +16,24 @@ class DartDependencyPackage {
   });
 
   factory DartDependencyPackage.fromHashMap(dynamic packageInfo) {
-    String name = '';
-    String version = '';
-    String source = '';
-    List<dynamic> dependencies = <dynamic>[];
+    List<String> dependencies = <String>[];
 
-    if (packageInfo is LinkedHashMap) {
-      final LinkedHashMap<String, dynamic> info = packageInfo as LinkedHashMap<String, dynamic>;
-      if (info.containsKey('name')) {
-        name = info['name'] as String;
-      }
-      if (info.containsKey('version')) {
-        version = info['version'] as String;
-      }
-      if (info.containsKey('source')) {
-        source = info['source'] as String;
-      }
-      if (info.containsKey('dependencies')) {
-        dependencies = info['dependencies'] as List<dynamic>;
-      }
+    if (packageInfo is! LinkedHashMap) {
+      return DartDependencyPackage(
+        name: '',
+        version: '',
+        source: '',
+        dependencies: dependencies,
+      );
+    }
+    if (packageInfo case {'dependencies': final List<Object?> list}) {
+      dependencies = <String>[for (final Object? e in list) '$e'];
```

Comment:
In case it helps, I'll copy-paste the diffs below so we have a version where GitHub isn't trying to line them up:

<details> <summary><h4>before</h4></summary>

```dart
  factory DartDependencyPackage.fromHashMap(dynamic packageInfo) {
    String name = '';
    String version = '';
    String source = '';
    List<dynamic> dependencies = <dynamic>[];

    if (packageInfo is LinkedHashMap) {
      final LinkedHashMap<String, dynamic> info = packageInfo as LinkedHashMap<String, dynamic>;
      if (info.containsKey('name')) {
        name = info['name'] as String;
      }
      if (info.containsKey('version')) {
        version = info['version'] as String;
      }
      if (info.containsKey('source')) {
        source = info['source'] as String;
      }
      if (info.containsKey('dependencies')) {
        dependencies = info['dependencies'] as List<dynamic>;
      }
    }
    return DartDependencyPackage(
      name: name,
      version: version,
      source: source,
      dependencies: dependencies.map((dynamic e) => e.toString()).toList(),
    );
  }
```

</details>
<details> <summary><h4>after</h4></summary>

```dart
  factory DartDependencyPackage.fromHashMap(dynamic packageInfo) {
    List<String> dependencies = <String>[];

    if (packageInfo is! LinkedHashMap) {
      return DartDependencyPackage(
        name: '',
        version: '',
        source: '',
        dependencies: dependencies,
      );
    }
    if (packageInfo case {'dependencies': final List<Object?> list}) {
      dependencies = <String>[for (final Object? e in list) '$e'];
    }
    return DartDependencyPackage(
      name: packageInfo['name'] as String? ?? '',
      version: packageInfo['version'] as String? ?? '',
      source: packageInfo['source'] as String? ?? '',
      dependencies: dependencies,
    );
  }
```

</details>

Reply:
Yes I have "split diff" on for github so I can see the old code on the left and new code on the right and I think the old code is easier to follow.

Reply:
For whatever it's worth I like this change, though I think it can be further improved by moving the declaration of `dependencies` to be after the `if (packageInfo is! LinkedHashMap)` branch.

Reply:
Also note that this introduces a change in behavior and is not only a refactor.

Before, we asserted that `info['dependencies']` is a `List` if it is non-null. For example if `info['dependencies']` were a `Map`, the code would crash. Now, `dependencies` would be left as an empty list.

Reply:
updated code:

```dart
if (packageInfo is! LinkedHashMap) {
  return DartDependencyPackage(
    name: '',
    version: '',
    source: '',
    dependencies: <String>[],
  );
}

return DartDependencyPackage(
  name: packageInfo['name'] as String? ?? '',
  version: packageInfo['version'] as String? ?? '',
  source: packageInfo['source'] as String? ?? '',
  dependencies: switch (packageInfo['dependencies']) {
    final List<Object?> list => list.map((Object? e) => '$e').toList(),
    _ => <String>[],
  },
);
```

I do like the simplification down to 2 `return` statements 🙂

Reply:
Apologies for all the force-pushing; I'm trying to keep commit labels organized.

> Before, we asserted that `info['dependencies']` is a `List` if it is non-null. For example if `info['dependencies']` were a `Map`, the code would crash. Now, `dependencies` would be left as an empty list.

good point, should be fixed in 9e715b840572074c94ac453d133a3238210b2096

Reply:
Indeed this is a lovely simplification; bit—to continue upon my [previous comment](https://github.com/flutter/flutter/pull/154753#discussion_r1785132313)—I think doing null-checks this way introduces a behavior change and a new class of bugs that would be very difficult to discover. Said another way, using `_ => <expression>` is an anti-pattern to me unless `<expression>` is a `throw` statement.

The last return statement could be rewritten as

```dart
// Note that I did not test this code locally. I hope this is all valid syntax.
return DartDependencyPackage(
  name: packageInfo['name'] as String? ?? '',
  version: packageInfo['version'] as String? ?? '',
  source: packageInfo['source'] as String? ?? '',
  dependencies: switch (packageInfo['dependencies']) {
    final List<Object?> list => list.map((Object? e) => '$e').toList(),
    null => <String>[],
    _ => throw Exception('dependencies field on <whatever packageInfo represents> has an expected type: ${packageInfo['dependencies'].runtimeType}',
  },
);
```

but then I would begin to argue that this is needless complex for most readers, and that pattern matching is no longer a good fit for use here.

IMO this is not a small issue and IIRC this particular change isn't the only instance of this problem, so I'll probably write a standalone comment about this.

Reply:
Sorry, this comment was written before https://github.com/flutter/flutter/pull/154753#discussion_r1785149186.

I do think [this](https://github.com/flutter/flutter/commit/9e715b840572074c94ac453d133a3238210b2096) addresses this sufficiently in a way I didn't think of 👍

Reply:
I am happy with the result here and think it is more readable than what we had before, even if I did have to look up `is!` because I thought it might be like `variable!`.